kore/src/Kore/Step/Simplification/Overloading.hs

```diff
@@ -239,8 +295,160 @@ unifyOverloadingVsOverloaded
+    vs.
+    inj{S2,injTo}(secondVariable)
+  If there exists a (unique) maximum overloadedHead' with its result sort
+  S2' within S2, such that there exists a common overload headUnion such that
+    inj{S1,injTo}(firstHead(firstChildren)) =
+        inj{S,injTo}(headUnion(inj1(firstChildren)))
+  and
+    inj{S2',injTo}(secondHead(freshVars)) =
+        inj{S,injTo}(headUnion(inj2(freshVars)))
+
+  Then it narrows overloadedVariable to inj{S2',S2}(secondHead(freshVars))
+  and reduces the problem to
+    inj{S,injTo}(headUnion(inj1(firstChildren)))
+   vs
+    inj{S,injTo}(headUnion(inj2(freshVars)))
+-}
+unifyOverloadingInjVsVariable
+    :: MonadSimplify unifier
+    => InternalVariable variable
+    => Application Symbol (TermLike variable)
+    -> ElementVariable variable
+    -> Attribute.FreeVariables variable
+    -> Inj ()
+    -> UnifyOverloadingResult unifier variable
+unifyOverloadingInjVsVariable
+    (Application firstHead firstChildren)
+    overloadedVar
+    freeVars
+    injProto
+  = do
+    OverloadSimplifier
+        { isOverloaded, resolveOverloading, unifyOverloadWithSortWithinBound }
+        <- Simplifier.askOverloadSimplifier
+    Monad.guard (isOverloaded firstHead)
+    case unifyOverloadWithSortWithinBound firstHead injProto of
+        Left err -> error err
+        Right Nothing -> notUnifiableOverloads
+        Right (Just (Pair (headUnion, maybeInjUnion) (overHead, maybeRInj))) ->
+            do
+            narrowing@Narrowing { narrowedPair = Pair _ second'}
+                <- computeNarrowing
+                    headUnion
+                    injProto
+                    freeVars
+                    overloadedVar
+                    overHead
+                    maybeRInj
+            let first' = resolveOverloading injProto headUnion firstChildren
+                inject = maybeMkInj maybeInjUnion
+            return $ WithNarrowing narrowing
+                { narrowedPair = Pair (inject first') (inject second') }
+
+computeNarrowing
+    :: MonadSimplify unifier
+    => InternalVariable variable
+    => Symbol -- ^overloading symbol
+    -> Inj () -- ^injection to overloading symbol
+    -> Attribute.FreeVariables variable -- ^free vars in the unification pair
+    -> ElementVariable variable -- ^injected variable (to be narrowed)
+    -> Symbol -- ^overloaded symbol computed for the variable
+    -> Maybe (Inj ()) -- ^injection into the variable's sort (if needed)
+    -> ExceptT UnifyOverloadingError unifier (Narrowing variable)
+computeNarrowing headUnion injUnion freeVars overloadedVar overHead maybeInj
+  = do
+    OverloadSimplifier { resolveOverloading }
+        <- Simplifier.askOverloadSimplifier
+    let second' =
+            resolveOverloading injUnion headUnion freshTerms
+    return Narrowing
+        { narrowedSubst = Predicate.markSimplified $
+            Predicate.makeEqualsPredicate_
+                (mkElemVar overloadedVar)
+                (maybeMkInj maybeInj overloadedTerm)
```

Comment:
Why not a `Substitution`?

Reply:
I guess that's because we never use it as a substitution; if we cannot reach bottom, we conjunct this with the rest of the conjunction and wrap it into exists quantifiers.